historical_data_api_client/cli.py

```diff
@@ -0,0 +1,110 @@
+"""
+import json
+import logging
+import os
+from datetime import datetime
+from logging import getLogger
+
+import click
+from click.types import Choice
+from click_default_group import DefaultGroup
+from colorlog.colorlog import ColoredFormatter
+
+from d3a_api_client.utils import get_simulation_id_from_config_uuid, get_auth_headers_for_requests
+from historical_data_api_client.get_historical_data import get_historical_device_profiles
+from historical_data_api_client.utils import validate_correct_datetime_format, DATE_TIME_FORMAT
+
+log = getLogger(__name__)
+
+
+@click.group(name='d3a-api-client', cls=DefaultGroup, default='run', default_if_no_args=True,
+             context_settings={'max_content_width': 120})
+@click.option('-l', '--log-level', type=Choice(list(logging._nameToLevel.keys())), default='INFO',
+              show_default=True, help="Log level")
+def main(log_level):
+    handler = logging.StreamHandler()
+    handler.setLevel(log_level)
+    handler.setFormatter(
+        ColoredFormatter(
+            "%(log_color)s%(asctime)s.%(msecs)03d %(levelname)-8s (%(lineno)4d) %(name)-30s: "
+            "%(message)s%(reset)s",
+            datefmt="%H:%M:%S"
+        )
+    )
+    root_logger = logging.getLogger()
+    root_logger.setLevel(log_level)
+    root_logger.addHandler(handler)
+
+
+@main.command()
+@click.option('-u', '--username', type=str, help="D3A username", required=None)
+@click.option('-p', '--password', type=str, help="D3A password", required=None)
+@click.option('-c', '--configuration-id', type=str, help="configuration uuid", required=True)
+@click.option('-a', '--area-uuid', type=str, help="area uuid", required=True)
+@click.option('-d', '--domain-name', default="https://d3aweb.gridsingularity.com",
+              type=str, help="D3A domain URL")
+@click.option('-s', '--start-time', type=str, default=None,
+              help="start time (Format(YYYY-MM-DDThh:mm)")
+@click.option('-e', '--end-time', type=str, default=None,
+              help="end time (Format(YYYY-MM-DDThh:mm)")
+@click.option('-t', '--time-resolution-hours', type=int, default=1,
+              help="time resolution in hours")
+@click.option('-r', '--export-file-path', type=str, default=None,
+              help="path to export json file")
+def run(username, password, domain_name, configuration_id, area_uuid, start_time, end_time,
+        time_resolution_hours, export_file_path):
+    try:
+        if username is None:
+            if "API_CLIENT_USERNAME" not in os.environ:
+                raise Exception("Please provide a username either via cli or env variable")
+        else:
+            os.environ["API_CLIENT_USERNAME"] = username
+        if password is None:
+            if "API_CLIENT_PASSWORD" not in os.environ:
+                raise Exception("Please provide a password either via cli or env variable")
+        else:
+            os.environ["API_CLIENT_PASSWORD"] = password
```

Comment:
I would suggest to only fetch env variables at the entrypoint of the code (i.e. here in `cli.py`) and to just pass them around as parameters. This would make testing easier, and it would make the code clearer in my opinion.
At the moment, this code doesn't tell us where the username and password will be used. If this could be avoided, one possible refactoring could be:
```suggestion
        username = username or os.getenv("API_CLIENT_USERNAME")
        password = password or os.getenv("API_CLIENT_PASSWORD")
        if not username:
            raise Exception("Please provide a username either via cli or env variable")
        if not password:
            raise Exception("Please provide a password either via cli or env variable")
            
        # And later pass the variables to the methods that use them
```

Reply:
The reason for not assigning a variable with these values is that `retrieve_jwt_key_from_server` is only using the env variables for authentication. This way we dont pass around secret information and the user can also set the username and password in the environment without passing it int he cli (which is normally frowned up on AFIAK, but we added it also in the other api-client in order to convenience the user). 
So in order to except both ways of providing the username and password, we need to go my way.
Are you convinced yet? ;-)

Reply:
TL;DR: your solution is good and backwards-compatible for this PR. In general (and for new code) I prefer passing parameters around and avoid env variables.

---

I'm partly convinced by what you wrote 😄 I see how the change I proposed would not be backwards-compatible with the previous code (which uses env variables, as you said). So I agree that your implementation is the best way to proceed without changing the existing implementation.

On the other hand, I'm not convinced that using global env variables is the best way to go. Secret information is passed around in any case (either in the `os.environ` dict or as parameter) and that probably doesn't make a difference in terms of security. Actually, env variables are visible from anywhere in the code, while parameters are only visible in the function scope.
Moreover, env variables can clash: what if another package uses the same `API_CLIENT_USERNAME` key? What if some method calls `os.environ.pop("API_CLIENT_USERNAME")` (as an extreme example)?

In any case, you can still use env variables in both cases (so the user shouldn't necessarily pass them in the cli). In my idea, the user would set up the env variables and these would be read in the `cli.py` script. From that moment on, they would be passed around as parameters, making it more explicit.

```python
# cli.py
def main():
    username = os.getenv("USERNAME")
    password = os.getenv("PASSWORD")
    retrieve_jwt_key_from_server(username=username, password=password)
```

Reply:
Ok, thanks a lot! Your point with the variable clash actually came up when I started implementing, but I shrugged it off ;-) 
Actually, I will probably do like you suggest as I will probably move this to its own repo.(https://github.com/gridsingularity/d3a-api-client/pull/156#pullrequestreview-701868057)
But I like more input for this topic first before I go that way.

Reply:
> Actually, I will probably do like you suggest as I will probably move this to its own repo.(#156 (review))

I would be in favour of this. If you go that way, I think it would be very nice to use [`python-dotenv`](https://github.com/theskumar/python-dotenv) to allow easier management of env variables.